File: problems/07/valid_anagram.c
#include <criterion/criterion.h>
#include <string.h>

int valid_anagram(const char *s, const char *t)
{
  int counts[26] = {0};

  if (strlen(s) != strlen(t))
  {
    return 0;
  }

  for (int i = 0; s[i] != '\0'; i++)
  {
    counts[s[i] - 'a']++;
    counts[t[i] - 'a']--;
  }

  // all values should be 0
  for (int i = 0; i < 26; i++)
  {
    if (counts[i])
    {
      return 0;
    }
  }

  return 1;
}

Test(ValidAnagram, LeetcodeExamples)
{
  cr_expect(valid_anagram("anagram", "nagaram") == 1);
  cr_expect(valid_anagram("rat", "tag") == 0);
}

File: problems/07/valid_anagram.exs
defmodule ValidAnagram do
  def solve(first, second) do
    count(first) == count(second)
  end

  def count(s) do
    count(String.to_charlist(s), %{})
  end

  def count([], acc), do: acc

  def count([ch | rest], acc) do
    case Map.fetch(acc, ch) do
      {:ok, cnt} -> count(rest, Map.put(acc, ch, cnt + 1))
      :error -> count(rest, Map.put(acc, ch, 1))
    end
  end
end

defmodule ValidAnagramTest do
  use ExUnit.Case

  test "leetcode tests" do
    assert ValidAnagram.solve("anagram", "nagaram") == true
    assert ValidAnagram.solve("rat", "car") == false
  end
end
